// react-rocketjump/lib/plugins/debounce/index.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _objectSpread = require('@babel/runtime/helpers/objectSpread2');
var rxjs = require('rxjs');
var operators = require('rxjs/operators');
var rjPlugin = require('../../rjPlugin.cjs.js');
require('@babel/runtime/helpers/slicedToArray');
require('@babel/runtime/helpers/toConsumableArray');
require('@babel/runtime/helpers/objectWithoutProperties');
require('@babel/runtime/helpers/toArray');
require('@babel/runtime/helpers/typeof');
require('object-path-immutable');
require('@babel/runtime/helpers/defineProperty');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _objectSpread__default = /*#__PURE__*/_interopDefaultLegacy(_objectSpread);

var defaultOptions = {
  time: 180
};

var rjDebounce = function rjDebounce() {
  var optionsOrTime = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : defaultOptions;
  var options;

  if (typeof optionsOrTime === 'number' && isFinite(optionsOrTime)) {
    options = {
      time: optionsOrTime
    };
  } else {
    options = _objectSpread__default['default'](_objectSpread__default['default']({}, defaultOptions), optionsOrTime);
  }

  return rjPlugin.rjPlugin({
    actions: function actions(_ref) {
      var run = _ref.run;
      return {
        runDebounced: function runDebounced() {
          return run.apply(void 0, arguments).withMeta({
            debounced: true
          });
        }
      };
    },
    effectPipeline: function effectPipeline(actionObservable) {
      return actionObservable.pipe(operators.scan(function (prev, current) {
        // Shoud debounce ma run?
        if (current.type === rjPlugin.RUN && current.meta && current.meta.debounced && typeof options.when === 'function') {
          var shouldDebounce = options.when(prev ? prev.payload.params : null, current.payload.params);

          var newMeta = _objectSpread__default['default']({}, current.meta);

          if (shouldDebounce) {
            newMeta.debounced = true;
          } else {
            newMeta.debounced = false;
          }

          return _objectSpread__default['default'](_objectSpread__default['default']({}, current), {}, {
            meta: newMeta
          });
        }

        return current;
      }, null), // The scan signature T
      operators.debounce(function (action) {
        if (action.type === rjPlugin.RUN && action.meta.debounced) {
          return rxjs.timer(options.time);
        } else {
          return rxjs.EMPTY;
        }
      }), operators.distinctUntilChanged(function (prevAction, currAction) {
        // not same stuff
        if (currAction.type !== prevAction.type) {
          return false;
        } // Ignore not debounced actions...


        if (currAction.type !== rjPlugin.RUN || !currAction.meta.debounced) {
          return false;
        } // cant be the same


        if (currAction.payload.params.length !== prevAction.payload.params.length) {
          return false;
        } // TODO: Maybe deep compare at 1 level depth of plain objects
        // compare params


        for (var i = 0; i < currAction.payload.params.length; i++) {
          if (currAction.payload.params[i] !== prevAction.payload.params[i]) {
            return false;
          }
        } // the same action =)


        return true;
      }));
    }
  });
};

exports.default = rjDebounce;
