// react-rocketjump/lib/plugins/listUpdate/index.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var rjPlugin = require('../../rjPlugin.cjs.js');
var objectPathImmutable = require('object-path-immutable');
require('@babel/runtime/helpers/slicedToArray');
require('@babel/runtime/helpers/toConsumableArray');
require('@babel/runtime/helpers/objectSpread2');
require('@babel/runtime/helpers/objectWithoutProperties');
require('@babel/runtime/helpers/toArray');
require('rxjs');
require('@babel/runtime/helpers/typeof');
require('@babel/runtime/helpers/defineProperty');
require('rxjs/operators');

var TYPE = 'RJ_LIST_UPDATE';

var defaultIdentity = function defaultIdentity(action, listItem) {
  return action.item.id === listItem.id;
};

var defaultUpdater = function defaultUpdater(action) {
  return action.item;
};

var rjListUpdate = function rjListUpdate() {
  var config = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  var identity = config.identity || defaultIdentity;
  var updater = config.updater || defaultUpdater;
  var path = config.path || 'data';
  return rjPlugin.rjPlugin({
    actions: function actions() {
      return {
        updateItem: function updateItem(item) {
          return {
            type: TYPE,
            item: item
          };
        }
      };
    },
    reducer: function reducer(oldReducer) {
      return function (state, action) {
        if (action.type === TYPE) {
          var list = rjPlugin.get(state, path);
          if (list) list = list.map(function (listItem) {
            return identity(action, listItem) ? updater(action, listItem) : listItem;
          });
          return objectPathImmutable.set(state, path, list);
        }

        return oldReducer(state, action);
      };
    }
  });
};

exports.TYPE = TYPE;
exports.default = rjListUpdate;
