// react-rocketjump/lib/plugins/listDelete/index.es.js
import { r as rjPlugin, g as get } from '../../rjPlugin.es.js';
import { set } from 'object-path-immutable';
import '@babel/runtime/helpers/esm/slicedToArray';
import '@babel/runtime/helpers/esm/toConsumableArray';
import '@babel/runtime/helpers/esm/objectSpread2';
import '@babel/runtime/helpers/esm/objectWithoutProperties';
import '@babel/runtime/helpers/esm/toArray';
import 'rxjs';
import '@babel/runtime/helpers/esm/typeof';
import '@babel/runtime/helpers/esm/defineProperty';
import 'rxjs/operators';

var TYPE = 'RJ_LIST_DELETE';

var defaultIdentity = function defaultIdentity(action, listItem) {
  return action.item.id === listItem.id;
};

var rjListDelete = function rjListDelete() {
  var config = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  var identity = config.identity || defaultIdentity;
  var path = config.path || 'data';
  return rjPlugin({
    actions: function actions() {
      return {
        deleteItem: function deleteItem(item) {
          return {
            type: TYPE,
            item: item
          };
        }
      };
    },
    reducer: function reducer(oldReducer) {
      return function (state, action) {
        if (action.type === TYPE) {
          if (process.env.NODE_ENV !== 'production') {
            if (config.warnPagination !== false) {
              var paginationPath = path.split('.').map(function (item, i, arr) {
                return i === arr.length - 1 ? 'pagination' : item;
              }).join('.');

              if (state && get(state, paginationPath)) {
                console.warn('It seems you are using this plugin on a paginated list. Remember that this plugin is agnostic wrt pagination, and will break it. To suppress this warning, set warnPagination: false in the config object');
              }
            }
          }

          var list = get(state, path);
          if (list) list = list.filter(function (listItem) {
            return !identity(action, listItem);
          });
          return set(state, path, list);
        }

        return oldReducer(state, action);
      };
    }
  });
};

export default rjListDelete;
export { TYPE };
