// react-rocketjump/lib/plugins/listDelete/index.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var rjPlugin = require('../../rjPlugin.cjs.js');
var objectPathImmutable = require('object-path-immutable');
require('@babel/runtime/helpers/slicedToArray');
require('@babel/runtime/helpers/toConsumableArray');
require('@babel/runtime/helpers/objectSpread2');
require('@babel/runtime/helpers/objectWithoutProperties');
require('@babel/runtime/helpers/toArray');
require('rxjs');
require('@babel/runtime/helpers/typeof');
require('@babel/runtime/helpers/defineProperty');
require('rxjs/operators');

var TYPE = 'RJ_LIST_DELETE';

var defaultIdentity = function defaultIdentity(action, listItem) {
  return action.item.id === listItem.id;
};

var rjListDelete = function rjListDelete() {
  var config = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  var identity = config.identity || defaultIdentity;
  var path = config.path || 'data';
  return rjPlugin.rjPlugin({
    actions: function actions() {
      return {
        deleteItem: function deleteItem(item) {
          return {
            type: TYPE,
            item: item
          };
        }
      };
    },
    reducer: function reducer(oldReducer) {
      return function (state, action) {
        if (action.type === TYPE) {
          if (process.env.NODE_ENV !== 'production') {
            if (config.warnPagination !== false) {
              var paginationPath = path.split('.').map(function (item, i, arr) {
                return i === arr.length - 1 ? 'pagination' : item;
              }).join('.');

              if (state && rjPlugin.get(state, paginationPath)) {
                console.warn('It seems you are using this plugin on a paginated list. Remember that this plugin is agnostic wrt pagination, and will break it. To suppress this warning, set warnPagination: false in the config object');
              }
            }
          }

          var list = rjPlugin.get(state, path);
          if (list) list = list.filter(function (listItem) {
            return !identity(action, listItem);
          });
          return objectPathImmutable.set(state, path, list);
        }

        return oldReducer(state, action);
      };
    }
  });
};

exports.TYPE = TYPE;
exports.default = rjListDelete;
