// react-rocketjump/lib/plugins/cache/index.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var rxjs = require('rxjs');
var operators = require('rxjs/operators');
var _toConsumableArray = require('@babel/runtime/helpers/toConsumableArray');
var _objectSpread = require('@babel/runtime/helpers/objectSpread2');
var _classCallCheck = require('@babel/runtime/helpers/classCallCheck');
var _createClass = require('@babel/runtime/helpers/createClass');
var rjPlugin = require('../../rjPlugin.cjs.js');
require('@babel/runtime/helpers/slicedToArray');
require('@babel/runtime/helpers/objectWithoutProperties');
require('@babel/runtime/helpers/toArray');
require('@babel/runtime/helpers/typeof');
require('object-path-immutable');
require('@babel/runtime/helpers/defineProperty');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _toConsumableArray__default = /*#__PURE__*/_interopDefaultLegacy(_toConsumableArray);
var _objectSpread__default = /*#__PURE__*/_interopDefaultLegacy(_objectSpread);
var _classCallCheck__default = /*#__PURE__*/_interopDefaultLegacy(_classCallCheck);
var _createClass__default = /*#__PURE__*/_interopDefaultLegacy(_createClass);

var LRUCache = /*#__PURE__*/function () {
  function LRUCache(namespace, size, store) {
    _classCallCheck__default['default'](this, LRUCache);

    this.namespace = namespace;
    this.size = size;
    this.store = store;
  }

  _createClass__default['default'](LRUCache, [{
    key: "_effectiveKey",
    value: function _effectiveKey(key) {
      return "".concat(this.namespace, "-").concat(key);
    }
  }, {
    key: "_meta",
    value: function _meta() {
      var m = this.store.getItem("$".concat(this.namespace, "-meta"));

      if (m) {
        return _objectSpread__default['default']({}, m);
      } else {
        return {
          count: 0,
          lru: [],
          dict: {}
        };
      }
    }
  }, {
    key: "_setMeta",
    value: function _setMeta(nextMeta) {
      this.store.setItem("$".concat(this.namespace, "-meta"), nextMeta);
    }
  }, {
    key: "_resetMeta",
    value: function _resetMeta() {
      this.store.removeItem("$".concat(this.namespace, "-meta"));
    }
  }, {
    key: "has",
    value: function has(key) {
      var realKey = this._effectiveKey(key);

      return !!this._meta().dict[realKey];
    }
  }, {
    key: "get",
    value: function get(key) {
      var realKey = this._effectiveKey(key);

      var meta = this._meta();

      if (!!meta.dict[realKey]) {
        meta.lru = [].concat(_toConsumableArray__default['default'](meta.lru.filter(function (key) {
          return key !== realKey;
        })), [realKey]);

        this._setMeta(meta);

        return this.store.getItem(realKey);
      } else {
        return undefined;
      }
    }
  }, {
    key: "set",
    value: function set(key, value) {
      var realKey = this._effectiveKey(key);

      var meta = this._meta();

      if (meta.count === this.size) {
        var keyToDelete = meta.lru.shift();
        this.store.removeItem(keyToDelete);
        delete meta.dict[keyToDelete];
        meta.count--;
      }

      meta.dict[realKey] = 1;
      meta.count++;
      meta.lru = [].concat(_toConsumableArray__default['default'](meta.lru), [realKey]);

      this._setMeta(meta);

      this.store.setItem(realKey, value);
    }
  }, {
    key: "clear",
    value: function clear() {
      var _this = this;

      var meta = this._meta();

      Object.keys(meta.dict).forEach(function (key) {
        _this.store.removeItem(key);
      });

      this._resetMeta();
    }
  }]);

  return LRUCache;
}();
var FIFOCache = /*#__PURE__*/function () {
  function FIFOCache(namespace, size, store) {
    _classCallCheck__default['default'](this, FIFOCache);

    this.namespace = namespace;
    this.size = size;
    this.store = store;
  }

  _createClass__default['default'](FIFOCache, [{
    key: "_effectiveKey",
    value: function _effectiveKey(key) {
      return "".concat(this.namespace, "-").concat(key);
    }
  }, {
    key: "_meta",
    value: function _meta() {
      var m = this.store.getItem("$".concat(this.namespace, "-meta"));

      if (m) {
        return _objectSpread__default['default']({}, m);
      } else {
        return {
          count: 0,
          queue: [],
          dict: {}
        };
      }
    }
  }, {
    key: "_setMeta",
    value: function _setMeta(nextMeta) {
      this.store.setItem("$".concat(this.namespace, "-meta"), nextMeta);
    }
  }, {
    key: "_resetMeta",
    value: function _resetMeta() {
      this.store.removeItem("$".concat(this.namespace, "-meta"));
    }
  }, {
    key: "has",
    value: function has(key) {
      var realKey = this._effectiveKey(key);

      return !!this._meta().dict[realKey];
    }
  }, {
    key: "get",
    value: function get(key) {
      var realKey = this._effectiveKey(key);

      var meta = this._meta();

      if (!!meta.dict[realKey]) {
        return this.store.getItem(realKey);
      } else {
        return undefined;
      }
    }
  }, {
    key: "set",
    value: function set(key, value) {
      var realKey = this._effectiveKey(key);

      var meta = this._meta();

      if (meta.count === this.size) {
        var keyToDelete = meta.queue.shift();
        this.store.removeItem(keyToDelete);
        delete meta.dict[keyToDelete];
        meta.count--;
      }

      meta.dict[realKey] = 1;
      meta.count++;
      meta.queue = [].concat(_toConsumableArray__default['default'](meta.queue), [realKey]);

      this._setMeta(meta);

      this.store.setItem(realKey, value);
    }
  }, {
    key: "clear",
    value: function clear() {
      var _this2 = this;

      var meta = this._meta();

      Object.keys(meta.dict).forEach(function (key) {
        _this2.store.removeItem(key);
      });

      this._resetMeta();
    }
  }]);

  return FIFOCache;
}();

var BACKEND = {};
var InMemoryStore = /*#__PURE__*/function () {
  function InMemoryStore() {
    _classCallCheck__default['default'](this, InMemoryStore);
  }

  _createClass__default['default'](InMemoryStore, [{
    key: "getItem",
    value: function getItem(key) {
      return BACKEND[key];
    }
  }, {
    key: "setItem",
    value: function setItem(key, value) {
      BACKEND[key] = value;
    }
  }, {
    key: "removeItem",
    value: function removeItem(key) {
      delete BACKEND[key];
    }
  }]);

  return InMemoryStore;
}();
var clearInMemoryStore = function clearInMemoryStore() {
  for (var k in BACKEND) {
    delete BACKEND[k];
  }
};
var SessionStorageStore = /*#__PURE__*/function () {
  function SessionStorageStore() {
    _classCallCheck__default['default'](this, SessionStorageStore);
  }

  _createClass__default['default'](SessionStorageStore, [{
    key: "getItem",
    value: function getItem(key) {
      var v = sessionStorage.getItem(key);

      if (v) {
        return JSON.parse(v);
      }

      return v;
    }
  }, {
    key: "setItem",
    value: function setItem(key, value) {
      sessionStorage.setItem(key, JSON.stringify(value));
    }
  }, {
    key: "removeItem",
    value: function removeItem(key) {
      sessionStorage.removeItem(key);
    }
  }]);

  return SessionStorageStore;
}();
var LocalStorageStore = /*#__PURE__*/function () {
  function LocalStorageStore() {
    _classCallCheck__default['default'](this, LocalStorageStore);
  }

  _createClass__default['default'](LocalStorageStore, [{
    key: "getItem",
    value: function getItem(key) {
      var v = localStorage.getItem(key);

      if (v) {
        return JSON.parse(v);
      }

      return v;
    }
  }, {
    key: "setItem",
    value: function setItem(key, value) {
      localStorage.setItem(key, JSON.stringify(value));
    }
  }, {
    key: "removeItem",
    value: function removeItem(key) {
      localStorage.removeItem(key);
    }
  }]);

  return LocalStorageStore;
}();

var defaultKey = function defaultKey() {
  for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
    args[_key] = arguments[_key];
  }

  return JSON.stringify(args);
};

var rjCache = function rjCache(config) {
  if (!config.ns) {
    throw new Error('RjCache requires the ns property to be set');
  }

  if (config.ns.includes('$')) {
    throw new Error('RjCache ns cannot contain the $ symbol');
  }

  if (!config.size) {
    throw new Error('RjCache needs a cache size to be defined');
  }

  var ns = config.ns;
  var size = config.size;
  var store = new (config.store || SessionStorageStore)();
  var provider = new (config.provider || LRUCache)(ns, size, store);
  var key = config.key || defaultKey;
  return rjPlugin.rjPlugin({
    actions: function actions() {
      return {
        resetCache: function resetCache() {
          return rjPlugin.makeEffectAction('$reset-cache');
        }
      };
    },
    effectCaller: function effectCaller(effectFn) {
      for (var _len2 = arguments.length, args = new Array(_len2 > 1 ? _len2 - 1 : 0), _key2 = 1; _key2 < _len2; _key2++) {
        args[_key2 - 1] = arguments[_key2];
      }

      var k = key.apply(void 0, args);

      if (provider.has(k)) {
        return rxjs.of(provider.get(k));
      } else {
        return rxjs.from(effectFn.apply(void 0, args)).pipe(operators.map(function (result) {
          provider.set(k, result);
          return result;
        }));
      }
    },
    effectPipeline: function effectPipeline(action$) {
      return action$.pipe(operators.tap(function (action) {
        if (action.type === '$reset-cache') {
          provider.clear();
        }
      }), operators.filter(function (action) {
        return action.type !== '$reset-cache';
      }));
    }
  });
};

exports.FIFOCache = FIFOCache;
exports.InMemoryStore = InMemoryStore;
exports.LRUCache = LRUCache;
exports.LocalStorageStore = LocalStorageStore;
exports.SessionStorageStore = SessionStorageStore;
exports.clearInMemoryStore = clearInMemoryStore;
exports.default = rjCache;
