// react-rocketjump/lib/plugins/cache/index.es.js
import { of, from } from 'rxjs';
import { map, tap, filter } from 'rxjs/operators';
import _toConsumableArray from '@babel/runtime/helpers/esm/toConsumableArray';
import _objectSpread from '@babel/runtime/helpers/esm/objectSpread2';
import _classCallCheck from '@babel/runtime/helpers/esm/classCallCheck';
import _createClass from '@babel/runtime/helpers/esm/createClass';
import { r as rjPlugin, b as makeEffectAction } from '../../rjPlugin.es.js';
import '@babel/runtime/helpers/esm/slicedToArray';
import '@babel/runtime/helpers/esm/objectWithoutProperties';
import '@babel/runtime/helpers/esm/toArray';
import '@babel/runtime/helpers/esm/typeof';
import 'object-path-immutable';
import '@babel/runtime/helpers/esm/defineProperty';

var LRUCache = /*#__PURE__*/function () {
  function LRUCache(namespace, size, store) {
    _classCallCheck(this, LRUCache);

    this.namespace = namespace;
    this.size = size;
    this.store = store;
  }

  _createClass(LRUCache, [{
    key: "_effectiveKey",
    value: function _effectiveKey(key) {
      return "".concat(this.namespace, "-").concat(key);
    }
  }, {
    key: "_meta",
    value: function _meta() {
      var m = this.store.getItem("$".concat(this.namespace, "-meta"));

      if (m) {
        return _objectSpread({}, m);
      } else {
        return {
          count: 0,
          lru: [],
          dict: {}
        };
      }
    }
  }, {
    key: "_setMeta",
    value: function _setMeta(nextMeta) {
      this.store.setItem("$".concat(this.namespace, "-meta"), nextMeta);
    }
  }, {
    key: "_resetMeta",
    value: function _resetMeta() {
      this.store.removeItem("$".concat(this.namespace, "-meta"));
    }
  }, {
    key: "has",
    value: function has(key) {
      var realKey = this._effectiveKey(key);

      return !!this._meta().dict[realKey];
    }
  }, {
    key: "get",
    value: function get(key) {
      var realKey = this._effectiveKey(key);

      var meta = this._meta();

      if (!!meta.dict[realKey]) {
        meta.lru = [].concat(_toConsumableArray(meta.lru.filter(function (key) {
          return key !== realKey;
        })), [realKey]);

        this._setMeta(meta);

        return this.store.getItem(realKey);
      } else {
        return undefined;
      }
    }
  }, {
    key: "set",
    value: function set(key, value) {
      var realKey = this._effectiveKey(key);

      var meta = this._meta();

      if (meta.count === this.size) {
        var keyToDelete = meta.lru.shift();
        this.store.removeItem(keyToDelete);
        delete meta.dict[keyToDelete];
        meta.count--;
      }

      meta.dict[realKey] = 1;
      meta.count++;
      meta.lru = [].concat(_toConsumableArray(meta.lru), [realKey]);

      this._setMeta(meta);

      this.store.setItem(realKey, value);
    }
  }, {
    key: "clear",
    value: function clear() {
      var _this = this;

      var meta = this._meta();

      Object.keys(meta.dict).forEach(function (key) {
        _this.store.removeItem(key);
      });

      this._resetMeta();
    }
  }]);

  return LRUCache;
}();
var FIFOCache = /*#__PURE__*/function () {
  function FIFOCache(namespace, size, store) {
    _classCallCheck(this, FIFOCache);

    this.namespace = namespace;
    this.size = size;
    this.store = store;
  }

  _createClass(FIFOCache, [{
    key: "_effectiveKey",
    value: function _effectiveKey(key) {
      return "".concat(this.namespace, "-").concat(key);
    }
  }, {
    key: "_meta",
    value: function _meta() {
      var m = this.store.getItem("$".concat(this.namespace, "-meta"));

      if (m) {
        return _objectSpread({}, m);
      } else {
        return {
          count: 0,
          queue: [],
          dict: {}
        };
      }
    }
  }, {
    key: "_setMeta",
    value: function _setMeta(nextMeta) {
      this.store.setItem("$".concat(this.namespace, "-meta"), nextMeta);
    }
  }, {
    key: "_resetMeta",
    value: function _resetMeta() {
      this.store.removeItem("$".concat(this.namespace, "-meta"));
    }
  }, {
    key: "has",
    value: function has(key) {
      var realKey = this._effectiveKey(key);

      return !!this._meta().dict[realKey];
    }
  }, {
    key: "get",
    value: function get(key) {
      var realKey = this._effectiveKey(key);

      var meta = this._meta();

      if (!!meta.dict[realKey]) {
        return this.store.getItem(realKey);
      } else {
        return undefined;
      }
    }
  }, {
    key: "set",
    value: function set(key, value) {
      var realKey = this._effectiveKey(key);

      var meta = this._meta();

      if (meta.count === this.size) {
        var keyToDelete = meta.queue.shift();
        this.store.removeItem(keyToDelete);
        delete meta.dict[keyToDelete];
        meta.count--;
      }

      meta.dict[realKey] = 1;
      meta.count++;
      meta.queue = [].concat(_toConsumableArray(meta.queue), [realKey]);

      this._setMeta(meta);

      this.store.setItem(realKey, value);
    }
  }, {
    key: "clear",
    value: function clear() {
      var _this2 = this;

      var meta = this._meta();

      Object.keys(meta.dict).forEach(function (key) {
        _this2.store.removeItem(key);
      });

      this._resetMeta();
    }
  }]);

  return FIFOCache;
}();

var BACKEND = {};
var InMemoryStore = /*#__PURE__*/function () {
  function InMemoryStore() {
    _classCallCheck(this, InMemoryStore);
  }

  _createClass(InMemoryStore, [{
    key: "getItem",
    value: function getItem(key) {
      return BACKEND[key];
    }
  }, {
    key: "setItem",
    value: function setItem(key, value) {
      BACKEND[key] = value;
    }
  }, {
    key: "removeItem",
    value: function removeItem(key) {
      delete BACKEND[key];
    }
  }]);

  return InMemoryStore;
}();
var clearInMemoryStore = function clearInMemoryStore() {
  for (var k in BACKEND) {
    delete BACKEND[k];
  }
};
var SessionStorageStore = /*#__PURE__*/function () {
  function SessionStorageStore() {
    _classCallCheck(this, SessionStorageStore);
  }

  _createClass(SessionStorageStore, [{
    key: "getItem",
    value: function getItem(key) {
      var v = sessionStorage.getItem(key);

      if (v) {
        return JSON.parse(v);
      }

      return v;
    }
  }, {
    key: "setItem",
    value: function setItem(key, value) {
      sessionStorage.setItem(key, JSON.stringify(value));
    }
  }, {
    key: "removeItem",
    value: function removeItem(key) {
      sessionStorage.removeItem(key);
    }
  }]);

  return SessionStorageStore;
}();
var LocalStorageStore = /*#__PURE__*/function () {
  function LocalStorageStore() {
    _classCallCheck(this, LocalStorageStore);
  }

  _createClass(LocalStorageStore, [{
    key: "getItem",
    value: function getItem(key) {
      var v = localStorage.getItem(key);

      if (v) {
        return JSON.parse(v);
      }

      return v;
    }
  }, {
    key: "setItem",
    value: function setItem(key, value) {
      localStorage.setItem(key, JSON.stringify(value));
    }
  }, {
    key: "removeItem",
    value: function removeItem(key) {
      localStorage.removeItem(key);
    }
  }]);

  return LocalStorageStore;
}();

var defaultKey = function defaultKey() {
  for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
    args[_key] = arguments[_key];
  }

  return JSON.stringify(args);
};

var rjCache = function rjCache(config) {
  if (!config.ns) {
    throw new Error('RjCache requires the ns property to be set');
  }

  if (config.ns.includes('$')) {
    throw new Error('RjCache ns cannot contain the $ symbol');
  }

  if (!config.size) {
    throw new Error('RjCache needs a cache size to be defined');
  }

  var ns = config.ns;
  var size = config.size;
  var store = new (config.store || SessionStorageStore)();
  var provider = new (config.provider || LRUCache)(ns, size, store);
  var key = config.key || defaultKey;
  return rjPlugin({
    actions: function actions() {
      return {
        resetCache: function resetCache() {
          return makeEffectAction('$reset-cache');
        }
      };
    },
    effectCaller: function effectCaller(effectFn) {
      for (var _len2 = arguments.length, args = new Array(_len2 > 1 ? _len2 - 1 : 0), _key2 = 1; _key2 < _len2; _key2++) {
        args[_key2 - 1] = arguments[_key2];
      }

      var k = key.apply(void 0, args);

      if (provider.has(k)) {
        return of(provider.get(k));
      } else {
        return from(effectFn.apply(void 0, args)).pipe(map(function (result) {
          provider.set(k, result);
          return result;
        }));
      }
    },
    effectPipeline: function effectPipeline(action$) {
      return action$.pipe(tap(function (action) {
        if (action.type === '$reset-cache') {
          provider.clear();
        }
      }), filter(function (action) {
        return action.type !== '$reset-cache';
      }));
    }
  });
};

export default rjCache;
export { FIFOCache, InMemoryStore, LRUCache, LocalStorageStore, SessionStorageStore, clearInMemoryStore };
