// react-rocketjump/lib/plugins/plainList/index.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _objectSpread = require('@babel/runtime/helpers/objectSpread2');
var plugins_listInsert_index = require('../listInsert/index.cjs.js');
var plugins_listUpdate_index = require('../listUpdate/index.cjs.js');
var plugins_listDelete_index = require('../listDelete/index.cjs.js');
var rjPlugin = require('../../rjPlugin.cjs.js');
require('object-path-immutable');
require('@babel/runtime/helpers/slicedToArray');
require('@babel/runtime/helpers/toConsumableArray');
require('@babel/runtime/helpers/objectWithoutProperties');
require('@babel/runtime/helpers/toArray');
require('rxjs');
require('@babel/runtime/helpers/typeof');
require('@babel/runtime/helpers/defineProperty');
require('rxjs/operators');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _objectSpread__default = /*#__PURE__*/_interopDefaultLegacy(_objectSpread);

var makeListDataReducer = function makeListDataReducer(customListReducer) {
  var defaultListReducer = function defaultListReducer(prevState, _ref) {
    var data = _ref.payload.data,
        meta = _ref.meta;
    var newList = data;

    if (meta && meta.append) {
      return (prevState || []).concat(newList);
    }

    if (meta && meta.prepend) {
      return newList.concat(prevState || []);
    }

    return newList;
  };

  var listReducer = typeof customListReducer === 'function' ? customListReducer : defaultListReducer;
  return function (prevState, action) {
    return listReducer(prevState, action);
  };
};
// Selectors for a list
var makeListSelectors = function makeListSelectors(getData) {
  var getList = getData;

  var getCount = function getCount(state) {
    var data = getList(state);
    return data === null ? null : data.length;
  };

  return {
    getList: getList,
    getCount: getCount
  };
};

function rjPlainList() {
  var config = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  var dataReducer = makeListDataReducer(config.customListReducer);
  return rjPlugin.rjPlugin(plugins_listInsert_index['default'](), plugins_listUpdate_index['default'](), plugins_listDelete_index['default'](), {
    selectors: function selectors(_ref2) {
      var getData = _ref2.getData;
      return makeListSelectors(getData);
    },
    reducer: function reducer(oldReducer) {
      return function (state, action) {
        if (action.type === rjPlugin.SUCCESS) {
          return _objectSpread__default['default'](_objectSpread__default['default']({}, state), {}, {
            pending: false,
            data: dataReducer(state.data, action)
          });
        } else {
          return oldReducer(state, action);
        }
      };
    }
  });
}

exports.default = rjPlainList;
exports.makeListDataReducer = makeListDataReducer;
exports.makeListSelectors = makeListSelectors;
