// react-rocketjump/lib/plugins/map/index.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _defineProperty = require('@babel/runtime/helpers/defineProperty');
var _objectSpread = require('@babel/runtime/helpers/objectSpread2');
var rjPlugin = require('../../rjPlugin.cjs.js');
require('@babel/runtime/helpers/slicedToArray');
require('@babel/runtime/helpers/toConsumableArray');
require('@babel/runtime/helpers/objectWithoutProperties');
require('@babel/runtime/helpers/toArray');
require('rxjs');
require('@babel/runtime/helpers/typeof');
require('object-path-immutable');
require('rxjs/operators');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _defineProperty__default = /*#__PURE__*/_interopDefaultLegacy(_defineProperty);
var _objectSpread__default = /*#__PURE__*/_interopDefaultLegacy(_objectSpread);

var defaultKeyMaker = function defaultKeyMaker(action) {
  return action.meta ? action.meta.id : null;
};

var defaultDataTransform = function defaultDataTransform(arg) {
  return arg;
};

var handlePendingItem = function handlePendingItem(prevState, action) {
  return _objectSpread__default['default'](_objectSpread__default['default']({}, prevState), {}, {
    error: null,
    pending: true,
    data: prevState && prevState.data || null
  });
};

var handleSuccessItem = function handleSuccessItem(prevState, action, dataTransform) {
  return _objectSpread__default['default'](_objectSpread__default['default']({}, prevState), {}, {
    pending: false,
    data: dataTransform(action.payload.data)
  });
};

var handleFailureItem = function handleFailureItem(prevState, action) {
  return _objectSpread__default['default'](_objectSpread__default['default']({}, prevState), {}, {
    pending: false,
    error: action.payload
  });
};

var makeMapReducer = function makeMapReducer(fallbackReducer) {
  var keyMaker = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : defaultKeyMaker;
  var dataTransform = arguments.length > 2 ? arguments[2] : undefined;
  var keepCompleted = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : true;
  return function (prevState, action) {
    var key = keyMaker(action);
    prevState = prevState || {};

    switch (action.type) {
      case rjPlugin.PENDING:
        {
          return _objectSpread__default['default'](_objectSpread__default['default']({}, prevState), {}, _defineProperty__default['default']({}, key, handlePendingItem(prevState[key])));
        }

      case rjPlugin.FAILURE:
        {
          return _objectSpread__default['default'](_objectSpread__default['default']({}, prevState), {}, _defineProperty__default['default']({}, key, handleFailureItem(prevState[key], action)));
        }

      case rjPlugin.SUCCESS:
        {
          if (keepCompleted) {
            return _objectSpread__default['default'](_objectSpread__default['default']({}, prevState), {}, _defineProperty__default['default']({}, key, handleSuccessItem(prevState[key], action, dataTransform || defaultDataTransform)));
          } else {
            return rjPlugin.omit(prevState, [key]);
          }
        }

      case rjPlugin.CLEAN:
        {
          // Clear key state
          if (key) {
            return rjPlugin.omit(prevState, [key]);
          } // Clear all the state


          return {};
        }

      default:
        return fallbackReducer(prevState, action);
    }
  };
};

var makeMapSelectors = function makeMapSelectors(_ref) {
  var getRoot = _ref.getRoot;

  var getMapPendings = function getMapPendings(state) {
    var rootState = getRoot(state);
    return Object.keys(rootState).reduce(function (r, key) {
      return rootState[key].pending ? _objectSpread__default['default'](_objectSpread__default['default']({}, r), {}, _defineProperty__default['default']({}, key, true)) : r;
    }, {});
  };

  var getMapLoadings = getMapPendings;

  var getMapFailures = function getMapFailures(state) {
    var rootState = getRoot(state);
    return Object.keys(rootState).reduce(function (r, key) {
      var error = rootState[key].error;
      return error !== null ? _objectSpread__default['default'](_objectSpread__default['default']({}, r), {}, _defineProperty__default['default']({}, key, error)) : r;
    }, {});
  };

  var getMapData = function getMapData(state) {
    var rootState = getRoot(state);
    return Object.keys(rootState).reduce(function (r, key) {
      var data = rootState[key].data;
      return data !== null ? _objectSpread__default['default'](_objectSpread__default['default']({}, r), {}, _defineProperty__default['default']({}, key, data)) : r;
    }, {});
  };

  return {
    getMapLoadings: getMapLoadings,
    getMapPendings: getMapPendings,
    getMapFailures: getMapFailures,
    getMapData: getMapData
  };
};

var rjMap = function rjMap() {
  var mapConfig = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  return rjPlugin.rjPlugin({
    actions: function actions(_ref2) {
      var run = _ref2.run,
          clean = _ref2.clean;
      return {
        runKey: function runKey(id) {
          for (var _len = arguments.length, params = new Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
            params[_key - 1] = arguments[_key];
          }

          return run.apply(void 0, [id].concat(params)).withMeta({
            id: id
          });
        },
        cleanKey: function cleanKey(id) {
          return clean(id).withMeta({
            id: id
          });
        }
      };
    },
    reducer: function reducer(oldReducer) {
      return makeMapReducer(oldReducer, mapConfig.key, mapConfig.dataTransform, mapConfig.keepCompleted);
    },
    selectors: makeMapSelectors,
    takeEffect: ['groupBy', typeof mapConfig.key === 'function' ? mapConfig.key : defaultKeyMaker]
  });
};

exports.default = rjMap;
