// react-rocketjump/lib/plugins/map/index.es.js
import _defineProperty from '@babel/runtime/helpers/esm/defineProperty';
import _objectSpread from '@babel/runtime/helpers/esm/objectSpread2';
import { r as rjPlugin, C as CLEAN, x as omit, S as SUCCESS, F as FAILURE, P as PENDING } from '../../rjPlugin.es.js';
import '@babel/runtime/helpers/esm/slicedToArray';
import '@babel/runtime/helpers/esm/toConsumableArray';
import '@babel/runtime/helpers/esm/objectWithoutProperties';
import '@babel/runtime/helpers/esm/toArray';
import 'rxjs';
import '@babel/runtime/helpers/esm/typeof';
import 'object-path-immutable';
import 'rxjs/operators';

var defaultKeyMaker = function defaultKeyMaker(action) {
  return action.meta ? action.meta.id : null;
};

var defaultDataTransform = function defaultDataTransform(arg) {
  return arg;
};

var handlePendingItem = function handlePendingItem(prevState, action) {
  return _objectSpread(_objectSpread({}, prevState), {}, {
    error: null,
    pending: true,
    data: prevState && prevState.data || null
  });
};

var handleSuccessItem = function handleSuccessItem(prevState, action, dataTransform) {
  return _objectSpread(_objectSpread({}, prevState), {}, {
    pending: false,
    data: dataTransform(action.payload.data)
  });
};

var handleFailureItem = function handleFailureItem(prevState, action) {
  return _objectSpread(_objectSpread({}, prevState), {}, {
    pending: false,
    error: action.payload
  });
};

var makeMapReducer = function makeMapReducer(fallbackReducer) {
  var keyMaker = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : defaultKeyMaker;
  var dataTransform = arguments.length > 2 ? arguments[2] : undefined;
  var keepCompleted = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : true;
  return function (prevState, action) {
    var key = keyMaker(action);
    prevState = prevState || {};

    switch (action.type) {
      case PENDING:
        {
          return _objectSpread(_objectSpread({}, prevState), {}, _defineProperty({}, key, handlePendingItem(prevState[key])));
        }

      case FAILURE:
        {
          return _objectSpread(_objectSpread({}, prevState), {}, _defineProperty({}, key, handleFailureItem(prevState[key], action)));
        }

      case SUCCESS:
        {
          if (keepCompleted) {
            return _objectSpread(_objectSpread({}, prevState), {}, _defineProperty({}, key, handleSuccessItem(prevState[key], action, dataTransform || defaultDataTransform)));
          } else {
            return omit(prevState, [key]);
          }
        }

      case CLEAN:
        {
          // Clear key state
          if (key) {
            return omit(prevState, [key]);
          } // Clear all the state


          return {};
        }

      default:
        return fallbackReducer(prevState, action);
    }
  };
};

var makeMapSelectors = function makeMapSelectors(_ref) {
  var getRoot = _ref.getRoot;

  var getMapPendings = function getMapPendings(state) {
    var rootState = getRoot(state);
    return Object.keys(rootState).reduce(function (r, key) {
      return rootState[key].pending ? _objectSpread(_objectSpread({}, r), {}, _defineProperty({}, key, true)) : r;
    }, {});
  };

  var getMapLoadings = getMapPendings;

  var getMapFailures = function getMapFailures(state) {
    var rootState = getRoot(state);
    return Object.keys(rootState).reduce(function (r, key) {
      var error = rootState[key].error;
      return error !== null ? _objectSpread(_objectSpread({}, r), {}, _defineProperty({}, key, error)) : r;
    }, {});
  };

  var getMapData = function getMapData(state) {
    var rootState = getRoot(state);
    return Object.keys(rootState).reduce(function (r, key) {
      var data = rootState[key].data;
      return data !== null ? _objectSpread(_objectSpread({}, r), {}, _defineProperty({}, key, data)) : r;
    }, {});
  };

  return {
    getMapLoadings: getMapLoadings,
    getMapPendings: getMapPendings,
    getMapFailures: getMapFailures,
    getMapData: getMapData
  };
};

var rjMap = function rjMap() {
  var mapConfig = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  return rjPlugin({
    actions: function actions(_ref2) {
      var run = _ref2.run,
          clean = _ref2.clean;
      return {
        runKey: function runKey(id) {
          for (var _len = arguments.length, params = new Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
            params[_key - 1] = arguments[_key];
          }

          return run.apply(void 0, [id].concat(params)).withMeta({
            id: id
          });
        },
        cleanKey: function cleanKey(id) {
          return clean(id).withMeta({
            id: id
          });
        }
      };
    },
    reducer: function reducer(oldReducer) {
      return makeMapReducer(oldReducer, mapConfig.key, mapConfig.dataTransform, mapConfig.keepCompleted);
    },
    selectors: makeMapSelectors,
    takeEffect: ['groupBy', typeof mapConfig.key === 'function' ? mapConfig.key : defaultKeyMaker]
  });
};

export default rjMap;
