// react-rocketjump/lib/plugins/mutationsPending/index.es.js
import { r as rjPlugin, s as matchMutationType, P as PENDING, S as SUCCESS, F as FAILURE } from '../../rjPlugin.es.js';
import '@babel/runtime/helpers/esm/slicedToArray';
import '@babel/runtime/helpers/esm/toConsumableArray';
import '@babel/runtime/helpers/esm/objectSpread2';
import '@babel/runtime/helpers/esm/objectWithoutProperties';
import '@babel/runtime/helpers/esm/toArray';
import 'rxjs';
import '@babel/runtime/helpers/esm/typeof';
import 'object-path-immutable';
import '@babel/runtime/helpers/esm/defineProperty';
import 'rxjs/operators';

function makeMutationsPendingReducer(trackTypes) {
  return function () {
    var state = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 0;
    var action = arguments.length > 1 ? arguments[1] : undefined;

    if (matchMutationType(action.type, trackTypes, PENDING)) {
      return state + 1;
    }

    if (matchMutationType(action.type, trackTypes, [SUCCESS, FAILURE])) {
      return state - 1;
    }

    return state;
  };
}

var rjMutationsPending = function rjMutationsPending(track) {
  return rjPlugin({
    combineReducers: {
      mutationsPending: makeMutationsPendingReducer(track !== null && track !== void 0 ? track : '*')
    },
    selectors: function selectors() {
      return {
        anyMutationPending: function anyMutationPending(state) {
          return state.mutationsPending > 0;
        }
      };
    }
  });
};

export default rjMutationsPending;
