// react-rocketjump/lib/plugins/mutationsPending/index.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var rjPlugin = require('../../rjPlugin.cjs.js');
require('@babel/runtime/helpers/slicedToArray');
require('@babel/runtime/helpers/toConsumableArray');
require('@babel/runtime/helpers/objectSpread2');
require('@babel/runtime/helpers/objectWithoutProperties');
require('@babel/runtime/helpers/toArray');
require('rxjs');
require('@babel/runtime/helpers/typeof');
require('object-path-immutable');
require('@babel/runtime/helpers/defineProperty');
require('rxjs/operators');

function makeMutationsPendingReducer(trackTypes) {
  return function () {
    var state = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 0;
    var action = arguments.length > 1 ? arguments[1] : undefined;

    if (rjPlugin.matchMutationType(action.type, trackTypes, rjPlugin.PENDING)) {
      return state + 1;
    }

    if (rjPlugin.matchMutationType(action.type, trackTypes, [rjPlugin.SUCCESS, rjPlugin.FAILURE])) {
      return state - 1;
    }

    return state;
  };
}

var rjMutationsPending = function rjMutationsPending(track) {
  return rjPlugin.rjPlugin({
    combineReducers: {
      mutationsPending: makeMutationsPendingReducer(track !== null && track !== void 0 ? track : '*')
    },
    selectors: function selectors() {
      return {
        anyMutationPending: function anyMutationPending(state) {
          return state.mutationsPending > 0;
        }
      };
    }
  });
};

exports.default = rjMutationsPending;
