// react-rocketjump/lib/plugins/listInsert/index.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var rjPlugin = require('../../rjPlugin.cjs.js');
var objectPathImmutable = require('object-path-immutable');
require('@babel/runtime/helpers/slicedToArray');
require('@babel/runtime/helpers/toConsumableArray');
require('@babel/runtime/helpers/objectSpread2');
require('@babel/runtime/helpers/objectWithoutProperties');
require('@babel/runtime/helpers/toArray');
require('rxjs');
require('@babel/runtime/helpers/typeof');
require('@babel/runtime/helpers/defineProperty');
require('rxjs/operators');

var TYPE = 'RJ_LIST_INSERT';

var defaultMerge = function defaultMerge(action, list) {
  return (list || []).concat([action.item]);
};

var rjListInsert = function rjListInsert() {
  var config = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  var path = config.path || 'data';
  return rjPlugin.rjPlugin({
    actions: function actions() {
      return {
        insertItem: function insertItem(item) {
          return {
            type: TYPE,
            item: item
          };
        }
      };
    },
    reducer: function reducer(oldReducer) {
      return function (state, action) {
        if (action.type === TYPE) {
          if (process.env.NODE_ENV !== 'production') {
            if (config.warnPagination !== false) {
              var paginationPath = path.split('.').map(function (item, i, arr) {
                return i === arr.length - 1 ? 'pagination' : item;
              }).join('.');

              if (state && rjPlugin.get(state, paginationPath)) {
                console.warn('It seems you are using this plugin on a paginated list. Remember that this plugin is agnostic wrt pagination, and will break it. To suppress this warning, set warnPagination: false in the config object');
              }
            }
          }

          var mergeFunc = config.merge ? config.merge : defaultMerge;
          return objectPathImmutable.set(state, path, mergeFunc(action, rjPlugin.get(state, path)));
        }

        return oldReducer(state, action);
      };
    }
  });
};

exports.TYPE = TYPE;
exports.default = rjListInsert;
